File: src/var.cc
// Copyright 2015 Google Inc. All rights reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// +build ignore

#include "var.h"

#include "eval.h"
#include "expr.h"
#include "log.h"
#include "strutil.h"

std::unordered_map<const Var*, std::string> Var::diagnostic_messages_;

const char* GetOriginStr(VarOrigin origin) {
  switch (origin) {
    case VarOrigin::UNDEFINED:
      return "undefined";
    case VarOrigin::DEFAULT:
      return "default";
    case VarOrigin::ENVIRONMENT:
      return "environment";
    case VarOrigin::ENVIRONMENT_OVERRIDE:
      return "environment override";
    case VarOrigin::FILE:
      return "file";
    case VarOrigin::COMMAND_LINE:
      return "command line";
    case VarOrigin::OVERRIDE:
      return "override";
    case VarOrigin::AUTOMATIC:
      return "automatic";
  }
  CHECK(false);
  return "*** broken origin ***";
}

Var::Var() : Var(VarOrigin::UNDEFINED, nullptr, Loc()) {}

Var::Var(VarOrigin origin, Frame* definition, Loc loc)
    : Evaluable(loc),
      definition_(definition),
      origin_(origin),
      readonly_(false),
      deprecated_(false),
      obsolete_(false),
      self_referential_(false) {}

Var::~Var() {
  diagnostic_messages_.erase(this);
}

void Var::AppendVar(Evaluator*, Value*) {
  CHECK(false);
}

void Var::SetDeprecated(const std::string_view& msg) {
  deprecated_ = true;
  diagnostic_messages_[this] = std::string(msg);
}

void Var::SetObsolete(const std::string_view& msg) {
  obsolete_ = true;
  diagnostic_messages_[this] = std::string(msg);
}

void Var::Used(Evaluator* ev, const Symbol& sym) const {
  if (obsolete_) {
    ev->Error(StringPrintf("*** %s is obsolete%s.", sym.c_str(),
                           diagnostic_message_text()));
  } else if (deprecated_) {
    WARN_LOC(ev->loc(), "%s has been deprecated%s.", sym.c_str(),
             diagnostic_message_text());
  }
}

const char* Var::diagnostic_message_text() const {
  auto it = diagnostic_messages_.find(this);
  return it == diagnostic_messages_.end() ? "" : it->second.c_str();
}

const std::string& Var::DeprecatedMessage() const {
  static const std::string empty_string;
  auto it = diagnostic_messages_.find(this);
  return it == diagnostic_messages_.end() ? empty_string : it->second;
}

Var* Var::Undefined() {
  static Var* undefined_var;
  if (!undefined_var) {
    undefined_var = new UndefinedVar();
  }
  return undefined_var;
}

SimpleVar::SimpleVar(VarOrigin origin, Frame* definition, Loc loc)
    : Var(origin, definition, loc) {}

SimpleVar::SimpleVar(const std::string& v,
                     VarOrigin origin,
                     Frame* definition,
                     Loc loc)
    : Var(origin, definition, loc), v_(v) {}

SimpleVar::SimpleVar(VarOrigin origin,
                     Frame* definition,
                     Loc loc,
                     Evaluator* ev,
                     Value* v)
    : Var(origin, definition, loc) {
  v->Eval(ev, &v_);
}

bool SimpleVar::IsFunc(Evaluator*) const {
  return false;
}

void SimpleVar::Eval(Evaluator* ev, std::string* s) const {
  ev->CheckStack();
  *s += v_;
}

void SimpleVar::AppendVar(Evaluator* ev, Value* v) {
  std::string buf;
  v->Eval(ev, &buf);
  v_.push_back(' ');
  v_ += buf;
  definition_ = ev->CurrentFrame();
}

std::string_view SimpleVar::String() const {
  return v_;
}

std::string SimpleVar::DebugString() const {
  return v_;
}

RecursiveVar::RecursiveVar(Value* v,
                           VarOrigin origin,
                           Frame* definition,
                           Loc loc,
                           std::string_view orig)
    : Var(origin, definition, loc), v_(v), orig_(orig) {}

bool RecursiveVar::IsFunc(Evaluator* ev) const {
  return v_->IsFunc(ev);
}

void RecursiveVar::Eval(Evaluator* ev, std::string* s) const {
  ev->CheckStack();
  v_->Eval(ev, s);
}

void RecursiveVar::AppendVar(Evaluator* ev, Value* v) {
  ev->CheckStack();
  v_ = Value::NewExpr(v->Location(), v_, Value::NewLiteral(" "), v);
  definition_ = ev->CurrentFrame();
}

void RecursiveVar::Used(Evaluator* ev, const Symbol& sym) const {
  if (SelfReferential()) {
    ERROR_LOC(
        Location(),
        StringPrintf(
            "*** Recursive variable \"%s\" references itself (eventually).",
            sym.c_str())
            .c_str());
  }

  Var::Used(ev, sym);
}

std::string_view RecursiveVar::String() const {
  return orig_;
}

std::string RecursiveVar::DebugString() const {
  return Value::DebugString(v_);
}

UndefinedVar::UndefinedVar() {}

bool UndefinedVar::IsFunc(Evaluator*) const {
  return false;
}

void UndefinedVar::Eval(Evaluator*, std::string*) const {
  // Nothing to do.
}

std::string_view UndefinedVar::String() const {
  return std::string_view("");
}

std::string UndefinedVar::DebugString() const {
  return "*undefined*";
}

VariableNamesVar::VariableNamesVar(std::string_view name, bool all)
    : name_(name), all_(all) {
  SetReadOnly();
  SetAssignOp(AssignOp::COLON_EQ);
}

bool VariableNamesVar::IsFunc(Evaluator*) const {
  return false;
}

void VariableNamesVar::Eval(Evaluator* ev, std::string* s) const {
  ConcatVariableNames(ev, s);
}

std::string_view VariableNamesVar::String() const {
  return name_;
}

void VariableNamesVar::ConcatVariableNames(Evaluator* ev,
                                           std::string* s) const {
  WordWriter ww(s);
  std::vector<std::string_view>&& symbols =
      GetSymbolNames([=](Var* var) -> bool {
        if (var->Obsolete()) {
          return false;
        }
        if (!all_) {
          if (var->IsFunc(ev)) {
            return false;
          }
        }
        return true;
      });
  for (auto entry : symbols) {
    ww.Write(entry);
  }
}

std::string VariableNamesVar::DebugString() const {
  return "*VariableNamesVar*";
}

bool ShellStatusVar::is_set_ = false;
int ShellStatusVar::shell_status_ = 0;
std::string ShellStatusVar::shell_status_string_;

ShellStatusVar::ShellStatusVar() {
  SetReadOnly();
  SetAssignOp(AssignOp::COLON_EQ);
}

void ShellStatusVar::SetValue(int newShellStatus) {
  if (!is_set_ || shell_status_ != newShellStatus) {
    shell_status_ = newShellStatus;
    is_set_ = true;
    shell_status_string_.clear();
  }
}

bool ShellStatusVar::IsDefined() const {
  return is_set_;
}

bool ShellStatusVar::IsFunc(Evaluator*) const {
  return false;
}

void ShellStatusVar::Eval(Evaluator* ev, std::string* s) const {
  if (ev->IsEvaluatingCommand()) {
    ev->Error("Kati does not support using .SHELLSTATUS inside of a rule");
  }

  if (!is_set_) {
    return;
  }

  *s += String();
}

std::string_view ShellStatusVar::String() const {
  if (!is_set_) {
    return "";
  }

  if (shell_status_string_.empty()) {
    shell_status_string_ = std::to_string(shell_status_);
  }

  return shell_status_string_;
}

std::string ShellStatusVar::DebugString() const {
  return "*ShellStatusVar*";
}

Vars::~Vars() {
  for (auto p : *this) {
    delete p.second;
  }
}

void Vars::add_used_env_vars(Symbol v) {
  used_env_vars_.insert(v);
}

Var* Vars::Lookup(Symbol name) const {
  auto found = find(name);
  if (found == end())
    return Var::Undefined();
  Var* v = found->second;
  if (v->Origin() == VarOrigin::ENVIRONMENT ||
      v->Origin() == VarOrigin::ENVIRONMENT_OVERRIDE) {
    used_env_vars_.insert(name);
  }
  return v;
}

Var* Vars::Peek(Symbol name) const {
  auto found = find(name);
  return found == end() ? Var::Undefined() : found->second;
}

void Vars::Assign(Symbol name, Var* v, bool* readonly) {
  *readonly = false;
  auto p = emplace(name, v);
  if (!p.second) {
    Var* orig = p.first->second;
    if (orig->ReadOnly()) {
      *readonly = true;
      return;
    }
    if (orig->Origin() == VarOrigin::OVERRIDE ||
        orig->Origin() == VarOrigin::ENVIRONMENT_OVERRIDE) {
      return;
    }
    if (orig->Origin() == VarOrigin::AUTOMATIC) {
      ERROR("overriding automatic variable is not implemented yet");
    }
    if (orig->IsDefined())
      delete p.first->second;
    p.first->second = v;
  }
}

SymbolSet Vars::used_env_vars_;

ScopedVar::ScopedVar(Vars* vars, Symbol name, Var* var)
    : vars_(vars), orig_(NULL) {
  auto p = vars->emplace(name, var);
  iter_ = p.first;
  if (!p.second) {
    orig_ = iter_->second;
    iter_->second = var;
  }
}

ScopedVar::~ScopedVar() {
  if (orig_) {
    iter_->second = orig_;
  } else {
    vars_->erase(iter_);
  }
}

File: src-rs/var.rs
/*
Copyright 2025 Google LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{
    borrow::Cow,
    collections::{HashMap, HashSet},
    ffi::OsString,
    fmt::Debug,
    os::unix::ffi::OsStrExt,
    sync::{Arc, LazyLock},
};

use anyhow::Result;
use bytes::{BufMut, Bytes};
use parking_lot::{Mutex, RwLock};

use crate::{
    command::AutoCommandVar,
    error, error_loc,
    eval::Frame,
    loc::Loc,
    strutil::{WordWriter, has_path_prefix},
    symtab::get_symbol_names,
    warn_loc,
};
use crate::{
    eval::Evaluator,
    expr::{Evaluable, Value},
    stmt::AssignOp,
    symtab::Symbol,
};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum VarOrigin {
    Default,
    Environment,
    EnvironmentOverride,
    File,
    CommandLine,
    Override,
    Automatic,
}

pub fn get_origin_str(origin: VarOrigin) -> &'static str {
    match origin {
        VarOrigin::Default => "default",
        VarOrigin::Environment => "environment",
        VarOrigin::EnvironmentOverride => "environment override",
        VarOrigin::File => "file",
        VarOrigin::CommandLine => "command line",
        VarOrigin::Override => "override",
        VarOrigin::Automatic => "automatic",
    }
}

pub type Var = Arc<RwLock<Variable>>;

#[derive(Debug)]
pub struct Variable {
    loc: Option<Loc>,

    definition: Option<Arc<Frame>>,

    origin: VarOrigin,

    pub assign_op: Option<AssignOp>,
    pub readonly: bool,
    pub deprecated: Option<Arc<String>>,
    obsolete: Option<Arc<String>>,

    visibility_prefix: Option<Vec<OsString>>,

    value: InnerVar,
}

#[derive(Debug)]
pub enum InnerVar {
    Simple(Vec<u8>),
    Recursive { v: Arc<Value>, orig: Bytes },
    AutoCommand(Symbol, AutoCommandVar),
    ShellStatus,
    VariableNames { name: Bytes, all: bool },
}

impl Variable {
    pub fn loc(&self) -> &Option<Loc> {
        &self.loc
    }
    pub fn origin(&self) -> VarOrigin {
        self.origin
    }
    pub fn definition(&self) -> &Option<Arc<Frame>> {
        &self.definition
    }
    pub fn obsolete(&self) -> bool {
        self.obsolete.is_some()
    }
    pub fn set_obsolete(&mut self, message: Arc<String>) {
        self.obsolete = Some(message);
    }
    pub fn flavor(&self) -> &'static str {
        match &self.value {
            InnerVar::Simple(_) => "simple",
            InnerVar::Recursive { .. } => "recursive",
            InnerVar::AutoCommand(_, _) => "undefined",
            InnerVar::ShellStatus => "simple",
            InnerVar::VariableNames { .. } => "kati_variable_names",
        }
    }
    pub fn used(&self, ev: &Evaluator, sym: &Symbol) -> Result<()> {
        if let Some(obsolete) = &self.obsolete {
            error_loc!(ev.loc.as_ref(), "*** {sym} is obsolete{obsolete}.");
        }
        if let Some(deprecated) = &self.deprecated {
            warn_loc!(ev.loc.as_ref(), "{sym} has been deprecated{deprecated}.");
        }
        Ok(())
    }
    pub fn set_visibility_prefix(&mut self, prefixes: Vec<OsString>, name: &Symbol) -> Result<()> {
        if self.visibility_prefix.is_none() {
            self.visibility_prefix = Some(prefixes);
        } else if self.visibility_prefix != Some(prefixes) {
            error!("Visibility prefix conflict on variable: {name}");
        }
        Ok(())
    }
    pub fn immediate_eval(&self) -> bool {
        matches!(&self.value, InnerVar::Simple(_))
    }
    pub fn append_var(
        &mut self,
        v: Arc<Value>,
        frame: Arc<Frame>,
        loc: Option<&Loc>,
    ) -> Result<()> {
        match &mut self.value {
            InnerVar::Simple(_) => {
                panic!("append_var should not be used when immediate_eval returns true")
            }
            InnerVar::Recursive { v: prev, .. } => {
                *prev = Arc::new(Value::List(
                    prev.loc(),
                    vec![
                        prev.clone(),
                        Arc::new(Value::Literal(None, Bytes::from_static(b" "))),
                        v,
                    ],
                ));
                self.definition = Some(frame);
            }
            InnerVar::AutoCommand(sym, _) => {
                error_loc!(loc, "appending to ${sym} is not supported");
            }
            InnerVar::ShellStatus => panic!(),
            InnerVar::VariableNames { .. } => panic!(),
        }
        Ok(())
    }
    pub fn append_str(&mut self, buf: &Bytes, frame: Arc<Frame>) -> Result<()> {
        match &mut self.value {
            InnerVar::Simple(s) => {
                s.push(b' ');
                s.extend_from_slice(buf);
                self.definition = Some(frame);
            }
            InnerVar::Recursive { v: prev, .. } => {
                *prev = Arc::new(Value::List(
                    prev.loc(),
                    vec![
                        prev.clone(),
                        Arc::new(Value::Literal(None, Bytes::from_static(b" "))),
                        Arc::new(Value::Literal(None, buf.clone())),
                    ],
                ));
                self.definition = Some(frame);
            }
            InnerVar::AutoCommand(sym, _) => {
                error!("appending to ${sym} is not supported");
            }
            InnerVar::ShellStatus => panic!(),
            InnerVar::VariableNames { .. } => panic!(),
        }
        Ok(())
    }
    pub fn check_current_referencing_file(&self, loc: &Option<Loc>, sym: Symbol) -> Result<()> {
        let Some(prefixes) = &self.visibility_prefix else {
            return Ok(());
        };
        let loc = loc.clone().unwrap_or_default();
        let mut valid = false;
        for prefix in prefixes {
            if has_path_prefix(&loc.filename.as_bytes(), prefix.as_bytes()) {
                valid = true;
                break;
            }
        }
        if !valid {
            let s = prefixes
                .iter()
                .map(|s| s.to_string_lossy())
                .collect::<Vec<Cow<str>>>()
                .join("\n");
            error!(
                "{} is not a valid file to reference variable {sym}. Line #{}.\nValid file prefixes:\n{s}",
                loc.filename, loc.line
            );
        }
        Ok(())
    }
    pub fn string(&self) -> Result<Cow<'_, [u8]>> {
        Ok(match &self.value {
            InnerVar::Simple(s) => Cow::Borrowed(s.as_slice()),
            InnerVar::Recursive { v: _, orig } => Cow::Borrowed(orig),
            InnerVar::AutoCommand(sym, _) => {
                error!("$(value {sym}) is not implemented yet");
            }
            InnerVar::ShellStatus => {
                Cow::Owned(if let Some(status) = SHELL_STATUS.lock().as_ref() {
                    status.to_string().as_bytes().to_vec()
                } else {
                    Vec::new()
                })
            }
            InnerVar::VariableNames { name, .. } => Cow::Borrowed(name),
        })
    }

    pub fn new_simple(
        origin: VarOrigin,
        frame: Option<Arc<Frame>>,
        loc: Option<Loc>,
    ) -> Arc<RwLock<Self>> {
        Arc::new(RwLock::new(Self {
            loc,
            definition: frame,
            origin,
            assign_op: None,
            readonly: false,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::Simple(Vec::new()),
        }))
    }

    pub fn with_simple_string(
        value: Bytes,
        origin: VarOrigin,
        frame: Option<Arc<Frame>>,
        loc: Option<Loc>,
    ) -> Arc<RwLock<Self>> {
        Arc::new(RwLock::new(Self {
            loc,
            definition: frame,
            origin,
            assign_op: None,
            readonly: false,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::Simple(value.to_vec()),
        }))
    }

    pub fn with_simple_value(
        origin: VarOrigin,
        frame: Option<Arc<Frame>>,
        loc: Option<Loc>,
        ev: &mut Evaluator,
        v: &Value,
    ) -> Result<Arc<RwLock<Self>>> {
        let value = v.eval_to_buf(ev)?;
        Ok(Arc::new(RwLock::new(Self {
            loc,
            definition: frame,
            origin,
            assign_op: None,
            readonly: false,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::Simple(value.to_vec()),
        })))
    }

    pub fn new_recursive(
        v: Arc<Value>,
        origin: VarOrigin,
        frame: Option<Arc<Frame>>,
        loc: Option<Loc>,
        orig: Bytes,
    ) -> Arc<RwLock<Self>> {
        Arc::new(RwLock::new(Self {
            loc,
            definition: frame,
            origin,
            assign_op: None,
            readonly: false,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::Recursive { v, orig },
        }))
    }

    pub fn new_autocommand(sym: Symbol, a: AutoCommandVar) -> Arc<RwLock<Self>> {
        Arc::new(RwLock::new(Self {
            loc: None,
            definition: None,
            origin: VarOrigin::Automatic,
            assign_op: None,
            readonly: false,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::AutoCommand(sym, a),
        }))
    }

    pub fn new_shell_status_var() -> Arc<RwLock<Self>> {
        Arc::new(RwLock::new(Self {
            loc: None,
            definition: None,
            origin: VarOrigin::Override,
            assign_op: Some(AssignOp::ColonEq),
            readonly: true,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::ShellStatus,
        }))
    }

    pub fn new_variable_names(name: &'static [u8], all: bool) -> Arc<RwLock<Self>> {
        Arc::new(RwLock::new(Self {
            loc: None,
            definition: None,
            origin: VarOrigin::Override,
            assign_op: Some(AssignOp::ColonEq),
            readonly: true,
            deprecated: None,
            obsolete: None,
            visibility_prefix: None,
            value: InnerVar::VariableNames {
                name: Bytes::from_static(name),
                all,
            },
        }))
    }
}

impl Evaluable for Variable {
    fn eval(&self, ev: &mut crate::eval::Evaluator, out: &mut dyn BufMut) -> Result<()> {
        match &self.value {
            InnerVar::Simple(v) => {
                out.put_slice(v);
            }
            InnerVar::Recursive { v, .. } => {
                v.eval(ev, out)?;
            }
            InnerVar::AutoCommand(_, a) => {
                a.eval(ev, out)?;
            }
            InnerVar::ShellStatus => {
                if ev.is_evaluating_command {
                    error_loc!(
                        ev.loc.as_ref(),
                        "Kati does not support using .SHELLSTATUS inside of a rule"
                    );
                }

                if let Some(status) = SHELL_STATUS.lock().as_ref() {
                    out.put_slice(format!("{status}").as_bytes());
                }
            }
            InnerVar::VariableNames { all, .. } => {
                let mut ww = WordWriter::new(out);
                let symbols = get_symbol_names(|var| {
                    if var.read().obsolete() {
                        return false;
                    }
                    true
                });
                for (sym, entry) in symbols {
                    if !*all
                        && let Some(var) = sym.peek_global_var()
                        && var.read().is_func()
                    {
                        continue;
                    }
                    ww.write(&entry);
                }
            }
        }
        Ok(())
    }
    fn is_func(&self) -> bool {
        match &self.value {
            InnerVar::Simple(_) => false,
            InnerVar::Recursive { v, .. } => v.is_func(),
            InnerVar::AutoCommand(_, _) => true,
            InnerVar::ShellStatus => false,
            InnerVar::VariableNames { .. } => false,
        }
    }
}

static SHELL_STATUS: LazyLock<Mutex<Option<i32>>> = LazyLock::new(|| Mutex::new(None));

pub fn set_shell_status_var(status: i32) {
    *SHELL_STATUS.lock() = Some(status)
}

pub static USED_ENV_VARS: LazyLock<Mutex<HashSet<Symbol>>> =
    LazyLock::new(|| Mutex::new(HashSet::new()));

pub struct Vars(pub Mutex<HashMap<Symbol, Var>>);

impl Default for Vars {
    fn default() -> Self {
        Self::new()
    }
}

impl Vars {
    pub fn new() -> Self {
        Vars(Mutex::new(HashMap::new()))
    }

    pub fn lookup(&self, sym: Symbol) -> Option<Var> {
        let ret = self.0.lock().get(&sym).cloned()?;
        match ret.read().origin() {
            VarOrigin::Environment | VarOrigin::EnvironmentOverride => {
                USED_ENV_VARS.lock().insert(sym);
            }
            _ => {}
        }
        Some(ret)
    }

    pub fn peek(&self, sym: Symbol) -> Option<Var> {
        self.0.lock().get(&sym).cloned()
    }

    pub fn assign(&self, sym: Symbol, var: Var, readonly: &mut bool) -> Result<()> {
        *readonly = false;
        let mut vars = self.0.lock();
        if let Some(orig) = vars.get_mut(&sym) {
            if orig.read().readonly {
                *readonly = true;
                return Ok(());
            }
            match orig.read().origin() {
                VarOrigin::Override | VarOrigin::EnvironmentOverride => return Ok(()),
                VarOrigin::Automatic => {
                    error!("overriding automatic variable is not implemented yet");
                }
                _ => {}
            }
            *orig = var;
        } else {
            vars.insert(sym, var);
        }
        Ok(())
    }

    pub fn merge_from(&self, vars: &Vars) {
        let mut to = self.0.lock();
        let from = vars.0.lock();

        for (sym, var) in from.iter() {
            to.insert(*sym, var.clone());
        }
    }
}

impl Clone for Vars {
    fn clone(&self) -> Self {
        let m = self.0.lock();
        Self(Mutex::new(m.clone()))
    }
}

impl Debug for Vars {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let vars = self.0.lock();
        f.debug_map().entries(vars.iter()).finish()
    }
}

pub struct ScopedVar {
    vars: Arc<Vars>,
    sym: Symbol,
    orig: Option<Var>,
}

impl ScopedVar {
    pub fn new(vars: Arc<Vars>, sym: Symbol, var: Var) -> Self {
        let orig = {
            let mut vars = vars.0.lock();
            vars.insert(sym, var)
        };
        Self { vars, sym, orig }
    }
}

impl Drop for ScopedVar {
    fn drop(&mut self) {
        let mut vars = self.vars.0.lock();
        if let Some(orig) = self.orig.clone() {
            vars.insert(self.sym, orig);
        } else {
            vars.remove(&self.sym);
        }
    }
}
